internal/pkg/describe/app.go

```diff
@@ -90,5 +100,39 @@ func NewAppDescriber(opt NewAppDescriberConfig) (*AppDescriber, error) {
 //
 // If the Version field does not exist, then it's a legacy template and it returns an deploy.LegacyAppTemplateVersion and nil error.
 func (d *AppDescriber) Version() (string, error) {
```

Comment:
Why do we do the minimum of the app template vs stackset template?  
Is there a particular reason why we don't have two separate methods (AppVersion vs. StackSetVersion)?

Reply:
This is because i don't want to confuse user by telling them their `AppVersion` is xxx and their `StackSetVersion` is xxx. For most of circumstances the app stack version and stackset version are the same.
```
func (d *AppDescriber) Version() (string, error)
```
And forget about the infrastructure, this function returns the version of the whole application template. If it is not up-to-date, we will try to upgrade the app (potentially we'll only upgrade one of their app stack or stackset). The existence of this version function is just to facilitate the log message telling them we are upgrading their app from version xxx to xxx, or skip upgrading.